lib/max_subarray.py

```diff
@@ -2,11 +2,21 @@
 def max_sub_array(nums):
     """ Returns the max subarray of the given list of numbers.
         Returns 0 if  nums is None or an empty list.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2)
+        Space Complexity: O(1)
     """
     if nums == None:
         return 0
     if len(nums) == 0:
         return 0
-    pass
+    max_overall = nums[0]
+    max_so_far = nums[0]
+    for index, num in enumerate(nums):
+        so_far_sum = 0
+        for each in nums[:index + 1]:
+            so_far_sum += each
+        max_so_far = max(so_far_sum, num)
```

Comment:
With your current solution, you calculate what the sum would be for values from index 0 to the current index,  and set 'max_so_far' to be either that calculated value or the current value you are iterating over. 

We can eliminate the need for that inner loop/calculation by instead just checking if `max_so_far` + the next value in the list is greater than the next value in the list. If `max_so_far + num` is greater than `num`, `max_so_far` in the next iteration will already hold the `so_far_sum` value you're currently calculating with a loop. 

If `max_so_far + num` is _less_ than `num`, that means the `so_far_sum`  shouldn't be a part of the subarray anyways, so there's no need to calculate it in this case either.

How would this change to your code affect your time and space complexity? 🤔

```suggestion
        max_so_far = max(max_so_far + num, num)
```